src/main/java/net/fabricmc/loader/metadata/V1ModMetadataParser.java

```diff
@@ -0,0 +1,615 @@
+				break;
+			case "custom":
+				V1ModMetadataParser.readCustomValues(reader, customValues);
+				break;
+			default:
+				reader.skipValue();
+				break;
+			}
+		}
+
+		// Finally close off the object
+		reader.endObject();
+
+		// Validate all required fields are resolved
+		if (id == null) {
+			throw new ParseMetadataException.MissingRequired("id");
+		}
+
+		if (version == null) {
+			throw new ParseMetadataException.MissingRequired("version");
+		}
+
+		return new V1ModMetadata(id, version, environment, entrypoints, jars, mixins, accessWidener, depends, recommends, suggests, conflicts, breaks, requires, name, description, authors, contributors, contact, license, icon, languageAdapters, customValues);
+	}
+
+	private static ModEnvironment readEnvironment(JsonReader reader) throws ParseMetadataException, IOException {
+		final String environment = reader.nextString().toLowerCase(Locale.ROOT);
+
+		if (environment.isEmpty() || environment.equals("*")) {
+			return ModEnvironment.UNIVERSAL;
+		} else if (environment.equals("client")) {
+			return ModEnvironment.CLIENT;
+		} else if (environment.equals("server")) {
+			return ModEnvironment.SERVER;
+		} else {
+			throw new ParseMetadataException("Invalid environment type: " + environment + "!", reader);
+		}
+	}
+
+	private static void readEntrypoints(JsonReader reader, Map<String, List<EntrypointMetadata>> entrypoints) throws IOException, ParseMetadataException {
+		// Entrypoints must be an object
+		if (reader.peek() != JsonToken.BEGIN_OBJECT) {
+			throw new ParseMetadataException("Entrypoints must be an object", reader);
+		}
+
+		reader.beginObject();
+
+		while (reader.hasNext()) {
+			final String key = reader.nextName();
+
+			List<EntrypointMetadata> metadata = new ArrayList<>();
+
+			if (reader.peek() != JsonToken.BEGIN_ARRAY) {
+				throw new ParseMetadataException("Entrypoint list must be an array!", reader);
+			}
+
+			reader.beginArray();
+
+			while (reader.hasNext()) {
+				String adapter = "default";
+				String value = null;
+
+				// Entrypoints may be specified directly as a string or as an object to allow specification of the language adapter to use.
+				switch (reader.peek()) {
+				case STRING:
+					value = reader.nextString();
+					break;
+				case BEGIN_OBJECT:
+					reader.beginObject();
+
+					while (reader.hasNext()) {
+						switch (reader.nextName()) {
+						case "adapter":
+							adapter = reader.nextString();
+							break;
+						case "value":
+							value = reader.nextString();
+							break;
+						default: // TODO: Ignore invalid elements in the entrypoint object?
+							reader.skipValue();
+							break;
+						}
+					}
+
+					reader.endObject();
+					break;
+				default:
+					throw new ParseMetadataException("Entrypoint must be a string or object with \"value\" field", reader);
+				}
+
+				if (value == null) {
+					throw new ParseMetadataException.MissingRequired("Entrypoint value must be present");
+				}
```

Comment:
I'd move the above test to where it can actually happen (in the object case above)